PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -154,62 +154,29 @@ static int __stdcall BrowseCallbackProc(HWND hwnd, UINT uMsg, LPARAM, LPARAM pDa
 
 generic_string folderBrowser(HWND parent, const generic_string & title, int outputCtrlID, const TCHAR *defaultStr)
 {
-	generic_string dirStr;
-
-	// This code was copied and slightly modifed from:
-	// http://www.bcbdev.com/faqs/faq62.htm
-
-	// SHBrowseForFolder returns a PIDL. The memory for the PIDL is
-	// allocated by the shell. Eventually, we will need to free this
-	// memory, so we need to get a pointer to the shell malloc COM
-	// object that will free the PIDL later on.
-	LPMALLOC pShellMalloc = 0;
-	if (::SHGetMalloc(&pShellMalloc) == NO_ERROR)
+	generic_string folderName;
+	CustomFileDialog dlg(parent);
+	dlg.setTitle(title.c_str());
+
+	// Get an initial directory from the edit control or from argument provided
+	TCHAR directory[MAX_PATH] = {};
+	if (outputCtrlID != 0)
+		::GetDlgItemText(parent, outputCtrlID, directory, _countof(directory));
+	directory[_countof(directory) - 1] = '\0';
+	if (!directory[0] && defaultStr)
+		dlg.setFolder(defaultStr);
+	else if (directory[0])
+		dlg.setFolder(directory);
+
+	const TCHAR* szDir = dlg.pickFolder();
+	if (szDir)
 	{
-		// If we were able to get the shell malloc object,
-		// then proceed by initializing the BROWSEINFO stuct
-		BROWSEINFO info;
-		memset(&info, 0, sizeof(info));
-		info.hwndOwner = parent;
-		info.pidlRoot = NULL;
-		TCHAR szDisplayName[MAX_PATH];
-		info.pszDisplayName = szDisplayName;
-		info.lpszTitle = title.c_str();
-		info.ulFlags = BIF_USENEWUI | BIF_NONEWFOLDERBUTTON;

```

Comment:
Do we really need to hide the "New folder" button?

Reply:
NO!